### PR TITLE
Adds Error Content Support

### LINKED ISSUE #19: Error Content Support
[AB#9707](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/9707)

### PR DESCRIPTION
**Pull request by Ndiritu**
Throws a `GraphServiceException` containing an error object for 4xx and 5xx responses

Closes #19

### CODE REVIEW

File: src/Http/GraphRequest.php [hunk 1]
```diff
@@ -101,19 +103,38 @@ public function __construct(string $requestType, string $endpoint, AbstractGraph
...
+     * Returns the final request URI after resolving $endpoint to base URL
+     *
+     * @return UriInterface
+     */
     public function getRequestUri(): UriInterface {
         return $this->requestUri;
     }
 
+    /**
+     * ReturnS the HTTP method used
```
[the remark this answers is not in the document]
Comment (Ndiritu) [reply src/Http/GraphRequest.php]: Addressed in latest commit

File: src/Exception/GraphServiceException.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+ * Class GraphServiceException
+ *
+ * Thrown when the Graph API returns 4xx or 5xx responses
+ *
+ * @package Microsoft\Graph\Exception
+ * @copyright 2021 Microsoft Corporation
+ * @license https://opensource.org/licenses/MIT MIT License
+ * @link https://developer.microsoft.com/graph
+ */
+class GraphServiceException extends GraphException
```
[the remark this answers is not in the document]
Comment (Ndiritu) [reply src/Exception/GraphServiceException.php:23]: Addressed in latest commit

File: src/Http/GraphError.php [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        }
+        return null;
+    }
+
+    /**
+     * Get the additional error info
+     *
+     * @return GraphError|null
+     */
+    public function getInnerError(): ?GraphError {
```
[the remark this answers is not in the document]
Comment (Ndiritu) [reply src/Http/GraphError.php]: Added in latest commit

File: src/Http/GraphRequest.php [hunk 4]
```diff
@@ -340,7 +371,7 @@ public function upload(string $path, ?ClientInterface $client = null)
     /**
      * Sets default headers based on baseUrl being a Graph endpoint or not
      */
-    private function initHeaders(string $baseUrl): void
+    private function initHeaders(): void
     {
         $coreSdkVersion = "graph-php-core/".GraphConstants::SDK_VERSION;
         $serviceLibSdkVersion = "Graph-php-".$this->graphClient->getSdkVersion();
```
[the remark this answers is not in the document]
Comment (Ndiritu) [reply src/Http/GraphRequest.php:377]: No worries. I've pinged Darrel asking whether we'll lose v1.x metrics on the if we make this change. Or whether we can accommodate both filters somehow on the charts
If he gives the greenlight I'll send a PR with this change.

[the remark this answers is not in the document]
Comment (Ndiritu) [reply src/Http/GraphRequest.php:377]: Confirmed that all will be good. The regex filters should handle this switch. Apparently DotNet made this switch as well.
Created a PR for this change

**Comment by ddyett** [comment]
please open issues to link these PRs to, so that work can be tracked.

**Comment by Ndiritu** [comment]
> please open issues to link these PRs to, so that work can be tracked.

The issues are currently on the v1 repo. Working with the team to transfer all Core issues to this repo & switch to tracking high level goals using ZenHub

**Comment by Ndiritu** [comment]
@MIchaelMainer updated this PR to throw native `InvalidArgumentException` for validation errors, `GraphClientException` for 4xx and `GraphServiceException` for 5xx

**Comment by Ndiritu** [comment]
> > please open issues to link these PRs to, so that work can be tracked.
> 
> The issues are currently on the v1 repo. Working with the team to transfer all Core issues to this repo & switch to tracking high level goals using ZenHub

@ddyett @shemogumbe Core issues have been transferred to this repo and linked to epics to the best of my knowledge (subject to review & change with the rest of the team)

**Review by MIchaelMainer — APPROVED** [review]
Conditionally approved per the answer to my question: This change brings up another consideration. GraphClientException::GraphResponseException is very appropriate for 4xx errors, how do we specify a client exception that does not represent a 4xx error response?

We may want something like:
GraphClientException::Exception -- exceptions that occur before an HTTP request is made, or more generally, applicable to how the client throws an exception independent of request/response. 
GraphClientResponseException::GraphResponseException -- 4xx exception.

**Comment by Ndiritu** [comment]
> Conditionally approved per the answer to my question: This change brings up another consideration. GraphClientException::GraphResponseException is very appropriate for 4xx errors, how do we specify a client exception that does not represent a 4xx error response?
> 
> We may want something like:
> GraphClientException::Exception -- exceptions that occur before an HTTP request is made, or more generally, applicable to how the client throws an exception independent of request/response.
> GraphClientResponseException::GraphResponseException -- 4xx exception.
> 
> I'd like to get your thoughts.

Unable to figure out a use-case for a non-4xx client exception that wouldn't be covered by a native exception type e.g. InvalidArgumentException for validation errors etc.

My understanding from the Teams thread was that all other exceptions would be native?

**Comment by MIchaelMainer** [comment]
That's fair.

**Review by MIchaelMainer — APPROVED** [review]

**Pull request merged**